libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h

```diff
@@ -0,0 +1,135 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "AP_RangeFinder.h"
+#include "AP_RangeFinder_Backend_Serial.h"
+
+#ifndef AP_RANGEFINDER_RDS02UF_ENABLED
+#define AP_RANGEFINDER_RDS02UF_ENABLED AP_RANGEFINDER_BACKEND_DEFAULT_ENABLED
+#endif
+
+#if AP_RANGEFINDER_RDS02UF_ENABLED
+
+#define RDS02_BUFFER_SIZE 50
+
+class AP_RangeFinder_RDS02UF : public AP_RangeFinder_Backend_Serial
+{
+
+public:
+
+    static AP_RangeFinder_Backend_Serial *create(
+        RangeFinder::RangeFinder_State &_state,
+        AP_RangeFinder_Params &_params) {
+        return new AP_RangeFinder_RDS02UF(_state, _params);
+    }
+
+protected:
+
+    MAV_DISTANCE_SENSOR _get_mav_distance_sensor_type() const override {
+        return MAV_DISTANCE_SENSOR_ULTRASOUND;
+    }
+
+
+private:
+    using AP_RangeFinder_Backend_Serial::AP_RangeFinder_Backend_Serial;
+
+    // Data Format for Benewake Rds02UF
+    // ===============================
+    // 21 bytes total per message:
+    // 1) 0x55
+    // 2) 0x55
+    // 3) address
+    // 4) error_code
+    // 5) FC_CODE_L (low 8bit)
+    // 6) FC_CODE_H (high 8bit)
+    // 7) LENGTH_L (low 8bit)
+    // 8) LENGTH_H (high 8bit)
+    // 9) REAL_DATA (10Byte)
+    // 10) CRC8
+    // 11) END_1 0xAA
+    // 12) END_2 0xAA
+    /// enum for handled messages
+    enum class RDS02UF_PARSE_STATE {
+        STATE0_SYNC_1 = 0,
+        STATE1_SYNC_2,
+        STATE2_ADDRESS,
+        STATE3_ERROR_CODE,
+        STATE4_FC_CODE_L,
+        STATE5_FC_CODE_H,
+        STATE6_LENGTH_L,
+        STATE7_LENGTH_H,
+        STATE8_REAL_DATA,
+        STATE9_CRC,
+        STATE10_END_1,
+        STATE11_END_2
+    };
+
+    RDS02UF_PARSE_STATE decode_state;
+    uint8_t parser_index;
+    uint8_t parser_buffer[RDS02_BUFFER_SIZE];
+    float _distance_m;
+
+    void ParseRds02(uint8_t *data_in, uint8_t datalengh);
+
+    // get a distance reading
+    bool get_reading(float &reading_m) override;
+
+    // get temperature reading in C.  returns true on success and populates temp argument
+    // bool get_temp(float &temp) const override;
+
+    uint16_t read_timeout_ms() const override { return 3000; }
+
+    uint8_t crc8(uint8_t *pbuf, int32_t len);
+
+    bool decode(uint8_t c);
+
+    const uint8_t crc8_table[256] = {
+    0x93,0x98,0xE4,0x46,0xEB,0xBA,0x04,0x4C,
```

Comment:
It would be far preferable to do the harder calculations to come up with the correct crc than looking it up in a table.  We really don't want to burn flash on this table if we can avoid it.